### PR TITLE
Add support for coverage at compile-time (CTFE)

### PR DESCRIPTION
**Pull request by nordlow**
Ready for review.

### CODE REVIEW

**Comment by nordlow** [comment]
Ready for review.

**Review by thewilsonator — APPROVED** [review]

File: source/dub/commandline.d
Comment (WebFreak001) [inline source/dub/commandline.d:1387]: new undocumented switch?

**Review by WebFreak001 — CHANGES_REQUESTED** [review]
missing the switch in the dmd compiler? (only added to LDC)

File: source/dub/commandline.d
Comment (nordlow) [reply source/dub/commandline.d:1387]: Indeed. Will fix. What about GDC?

Comment (WebFreak001) [reply source/dub/commandline.d:1387]: if there is a GDC switch, add it, otherwise it can be added later.

Comment (nordlow) [reply source/dub/commandline.d:1387]: Is there a GDC switch corresponding to dmd/ldc's -cov=ctfe, @ibuclaw?

Comment (WebFreak001) [reply source/dub/commandline.d:1387]: ok then no addition there yet. I would suggest to also remove this extra switch here, which doesn't seem to be a standard switch. (it does the same thing as `--build=unittest-cov-ctfe`)

**Comment by WebFreak001** [comment]
do you think we could add a basic test for this? (to see if coverage includes ctfe coverage / and code coverage?)

**Comment by nordlow** [comment]
> do you think we could add a basic test for this? (to see if coverage includes ctfe coverage / and code coverage?)

I can try. Does it suffice to add a file ending with .sh in `./test` for Github CI to run it automatically?

**Comment by nordlow** [comment]
> do you think we could add a basic test for this? (to see if coverage includes ctfe coverage / and code coverage?)

I've added a test now. Can you take a look at it?

**Comment by nordlow** [comment]
> > do you think we could add a basic test for this? (to see if coverage includes ctfe coverage / and code coverage?)
> 
> I've added a test at https://github.com/dlang/dub/pull/2256/files#diff-de1900a2e1b27d6b232d129bf4301eb15d387e92239ab0e9904e8ea3bf98bda6. Can you take a look at it, @WebFreak001?

I've added a test for unittest-cov-ctfe aswell now.

File: source/dub/commandline.d
Comment (WebFreak001) [reply source/dub/commandline.d:1387]: do you want to keep this switch or should we better remove it?

**Review by WebFreak001 — APPROVED** [review]
LGTM, do you want to keep the `--coverage-ctfe` switch or do you want to remove it? (it's basically only an alias to `--build=unittest-cov-ctfe` which imo we should avoid)

If you think it's better to keep it for consistency, we also need a PR to the dub-docs repo to document it.

**Comment by nordlow** [comment]
> LGTM, do you want to keep the `--coverage-ctfe` switch or do you want to remove it? (it's basically only an alias to `--build=unittest-cov-ctfe` which imo we should avoid)
> 
> If you think it's better to keep it for consistency, we also need a PR to the dub-docs repo to document it.

Let's keep it for consistency.

**Comment by WebFreak001** [comment]
dub-docs: https://github.com/dlang/dub-docs/

**Comment by nordlow** [comment]
> dub-docs: https://github.com/dlang/dub-docs/

Why aren't the dub-docs in the same repo?

**Comment by WebFreak001** [comment]
dub-docs is statically deployed through netlify, so documentation updates don't need to restart the registry and potentially invalidate session tokens or abort downloads when it's done more regularly.

**Comment by nordlow** [comment]
> dub-docs is statically deployed through netlify, so documentation updates don't need to restart the registry and potentially invalidate session tokens or abort downloads when it's done more regularly.

Ok to update the dub-docs after this is merged?

**Comment by WebFreak001** [comment]
we can merge this already when the CI passes, but having the dub-docs PR ready would be great.

**Comment by nordlow** [comment]
> we can merge this already when the CI passes, but having the dub-docs PR ready would be great.

Added https://github.com/dlang/dub-docs/pull/45.

**Comment by WebFreak001** [comment]
coverage test still seems to fail on linux right now

**Comment by nordlow** [comment]
> coverage test still seems to fail on linux right now

It's caused by

```
[INFO] Running /Users/runner/work/dub/dub/test/unittest-cov-ctfe/...
Performing "debug" build using dmd for x86_64.
test 1.0.0: building configuration "library"...
Target is a library. Skipping execution.
[ERROR] Run failure.
```

.

**Comment by nordlow** [comment]
For each job, every test is run 3 times with different settings every time. Why is that?

**Comment by nordlow** [comment]
> For each job, every test is run 3 times with different settings at least once leading to the .lst being in different locations which is causing the grep to fail. Why is that?

I realized I don't have to grep for contents in the .lst file. I can use `DFLAGS=-cov=100 dub ...` instead.

**Comment by nordlow** [comment]
> > For each job, every test is run 3 times with different settings at least once leading to the .lst being in different locations which is causing the grep to fail. Why is that?
> 
> I realized I don't have to grep for contents of the .lst file. I can use `DFLAGS=-cov=100 dub ...` instead.

And that...encounter a dub bug...

```sh
DUB=../bin/dub ./unittest-cov-ctfe.sh 
Generating test runner configuration 'test-test-library' for 'library' (library).
Performing "unittest-cov-ctfe" build using ldc2 for x86_64.
test 1.0.0: target for configuration "test-test-library" is up to date.
To force a rebuild of up-to-date targets, run again with --force.
Running unittest-cov-ctfe/test-test-library 
1 modules passed unittests
Error: unittest-cov-ctfe/.dub/code/test-test-library-unittest-cov-ctfe-linux.posix-x86_64-ldc_v1.30.0-beta1-918E82E1BCD342DB713295CD6B6740A3_dub_test_root.d is 0% covered, less than required 100%
Error: unittest-cov-ctfe/.dub/code/test-test-library-unittest-cov-ctfe-linux.posix-x86_64-ldc_v1.30.0-beta1-918E82E1BCD342DB713295CD6B6740A3_dub_test_root.d is 0% covered, less than required 100%
Program exited with code 1
```

. Is this a know issue, @WebFreak001 ?

**Comment by nordlow** [comment]
Ready for final review and merge, @WebFreak001!

**Comment by nordlow** [comment]
Accidentally pressed close.  Reopened.

**Comment by nordlow** [comment]
buildkite error https://buildkite.com/dlang/dub/builds/1864#01812ed4-8d7c-4ccb-9571-810b106f6b86 seems unrelated.

**Comment by nordlow** [comment]
Error in libmir/mir at https://buildkite.com/dlang/dub/builds/1864#01812ed4-8d7c-4ccb-9571-810b106f6b86/108-212 seems unrelated. Ping, @9il

**Comment by 9il** [comment]
It is. I wil update Mir

**Comment by nordlow** [comment]
> It is. I wil update Mir

Thank you.

**Comment by WebFreak001** [comment]
don't know if it's possible to restart buildkite here?

**Comment by nordlow** [comment]
Green now. Merge, @WebFreak001?

**Pull request merged**
